src/delay_node.cpp

```diff
@@ -0,0 +1,44 @@
+// Copyright 2022 AIT Austrian Institute of Technology GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <utility>
+#include "rclcpp/rclcpp.hpp"
+#include "topic_tools/delay_node.hpp"
+
+namespace topic_tools
+{
+using namespace std::chrono_literals; // NOLINT
+
+DelayNode::DelayNode(const rclcpp::NodeOptions & options)
+: ToolBaseNode("delay", options), delay_(0ns)
+{
+  delay_ = rclcpp::Duration::from_seconds(declare_parameter("delay", 0.0));
+  use_wall_clock_ = declare_parameter("use_wall_clock", false);
+}
+
+void DelayNode::process_message(std::shared_ptr<rclcpp::SerializedMessage> msg)
+{
+  auto now = use_wall_clock_ ? rclcpp::Clock{}.now() : this->now();
+  queue_.push_back({ now + delay_, msg });
+  while (!queue_.empty() && queue_.front().first <= now) {
```

Comment:
Maybe I don't understand this correctly, but this looks to me like it is doing this: "Every time we receive a message, add it to a queue with a publish time. If there are any messages in the queue that are past that publish time, publish them."

If that's correct, then this node is not correct. It will only publish messages on the schedule that it receives them, so at best it can delay "by one message", rather than "by an amount of time". If no new messages come in, then a delayed message will never be published, or if the rate of incoming messages is not _exactly_ aligned with the `delay_` then the timing will be wrong.

There are a few ways that you could approach building this correctly. One is creating one-off rclcpp TImers for each incoming message. Or, perhaps you could have a second republisher thread, watching a queue (protected with a condition_variable), and using `cv.wait_until` to do the timing on that thread. Something like that

Reply:
Thanks for the review! You're correct, I've only ever used this tool with continuously published messages and delays that align well with publish rates, so I haven't noticed. I will update the PR.